File: src/gsocialsim/kernel/stimulus_ingestion.py
from typing import Dict, List
from gsocialsim.stimuli.data_source import DataSource
from gsocialsim.stimuli.stimulus import Stimulus

class StimulusIngestionEngine:
    def __init__(self):
        self._data_sources: List[DataSource] = []
        self._stimuli_store: Dict[str, Stimulus] = {}

    def register_data_source(self, source: DataSource):
        self._data_sources.append(source)

    def get_stimulus(self, stimulus_id: str) -> Stimulus | None:
        return self._stimuli_store.get(stimulus_id)

    def tick(self, current_tick: int) -> List[Stimulus]:
        """ Polls all data sources and adds new stimuli to the world. """
        newly_added = []
        for source in self._data_sources:
            new_stimuli = source.get_stimuli(current_tick)
            for stimulus in new_stimuli:
                self._stimuli_store[stimulus.id] = stimulus
                newly_added.append(stimulus)
        return newly_added

File: cpp/src/stimulus_ingestion.cpp
#include "stimulus_ingestion.h"

void StimulusIngestionEngine::register_data_source(std::shared_ptr<DataSource> source) {
    if (!source) return;
    data_sources_.push_back(std::move(source));
}

const Stimulus* StimulusIngestionEngine::get_stimulus(const std::string& stimulus_id) const {
    auto it = stimuli_store_.find(stimulus_id);
    if (it == stimuli_store_.end()) return nullptr;
    return &it->second;
}

const std::unordered_map<std::string, Stimulus>& StimulusIngestionEngine::stimuli_store() const {
    return stimuli_store_;
}

std::vector<Stimulus> StimulusIngestionEngine::tick(int current_tick) {
    std::vector<Stimulus> newly_added;
    for (const auto& source : data_sources_) {
        if (!source) continue;
        auto stimuli = source->get_stimuli(current_tick);
        for (const auto& stimulus : stimuli) {
            stimuli_store_[stimulus.id] = stimulus;
            newly_added.push_back(stimulus);
        }
    }
    return newly_added;
}
